packages/cli/src/tools/config/types.flow.js

```diff
@@ -2,75 +2,222 @@
+ */
+type InquirerPromptT = any;
 
 /**
- * A map of hooks to run pre/post some of the CLI actions
+ * Settings that a library author can define in the configuration bundled with
+ * dependency for Android
+ *
+ * See UserDependencyConfigT for details
  */
-type HooksT = {
-  [key: string]: string,
-  prelink?: string,
-  postlink?: string,
+type DependencyParamsAndroidT = {
+  sourceDir?: string,
+  manifestPath?: string,
+  packageImportPath?: string,
+  packageInstance?: string,
 };
 
 /**
- * A map with additional platforms that ship with a dependency.
+ * Settings that user can define in the project configuration for Android
+ *
+ * See UserConfigT for details
  */
-export type PlatformsT = {
-  [key: string]: {
-    dependencyConfig?: Function,
-    projectConfig?: Function,
-    linkConfig?: Function,
-  },
+type ProjectParamsAndroidT = {
+  sourceDir?: string,
+  manifestPath?: string,
+  packageName?: string,
+  packageFolder?: string,
+  mainFilePath?: string,
+  stringsPath?: string,
+  settingsGradlePath?: string,
+  assetsPath?: string,
+  buildGradlePath?: string,
 };
 
-export type DependencyConfigT = {
-  dependency: {
-    platforms: {
-      android?: AndroidConfigParamsT,
-      ios?: IOSConfigParamsT,
-      [key: string]: any,
-    },
-    assets: string[],
-    hooks: HooksT,
-    params: InquirerPromptT[],
-  },
-  commands: string[],
-  platforms: PlatformsT,
+/**
+ * Settings that user can define in the project configuration for iOS.
+ * Same for dependency - we share the type.
+ *
+ * See UserDependencyConfigT and UserConfigT for details
+ */
+type ProjectParamsIOST = {
+  project?: string,
+  sharedLibraries?: string[],
+  libraryFolder?: string,
 };
 
-type _ProjectConfigT = {
+/**
+ * Final configuration object
+ */
+export type ConfigT = {|
+  // Root where the configuration has been resolved from
   root: string,
+
+  // Path to React Native source
+  reactNativePath: string,
+
+  // Object that contains configuration for a project (null, when platform not available)
+  project: {
+    android: ?ProjectConfigAndroidT,
+    ios: ?ProjectConfigIOST,
+    [key: string]: any,
+  },
+
+  // An array of assets as defined by the user
+  assets: string[],
+
+  // Map of the dependencies that are present in the project
   dependencies: {
     [key: string]: {
+      name: string,
       platforms: {
-        android: DependencyConfigAndroidT | null,
-        ios: DependencyConfigIOST | null,
+        android?: DependencyConfigAndroidT | null,
+        ios?: DependencyConfigIOST | null,
         [key: string]: any,
       },
       assets: string[],
-      hooks: HooksT,
+      hooks: {
+        [key: string]: string,
+        prelink?: string,
+        postlink?: string,
+      },
       params: InquirerPromptT[],
     },
   },
-  platforms: PlatformsT,
+
+  // Map of available platforms (built-ins and dynamically loaded)
+  platforms: {
+    [name: string]: any,
+    ios: {
```

Comment:
simplified this a bit to avoid duplication and copy-paste errors (see 3d65cda)

Reply:
actually used to be like that, I removed it for better readability. But whatever makes sense :)